File: source/src/test/java/org/rss/tools/mpl/ParserDelegateTest.java
package org.rss.tools.mpl;

import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.io.InputStream;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.rss.tools.mpl.domain.Document;
import org.rss.tools.mpl.global.ServiceProvider;
import org.rss.tools.mpl.parsing.MplDocumentParser;
import org.rss.tools.mpl.parsing.Parser;
import org.rss.tools.mpl.render.DocumentVisitor;

public class ParserDelegateTest {

	@BeforeAll
	public static void setup() {
		Main.registerDefaultProviders();
	}
	
	@Test
	void testDefault() throws Exception {
		DocumentVisitor render = ServiceProvider.get(DocumentVisitor.class);
		Parser parser = ServiceProvider.get(Parser.class);
		
		assertNotNull(render);
		assertTrue(parser instanceof MplDocumentParser);
		
		testOverrideParser();
	}
	
	// Needs to be executed in sequence. Otherwise it breaks the default test
//	@Test
	void testOverrideParser() throws Exception {
		ServiceProvider.register(Parser.class, new MyParser());
		Parser parser = ServiceProvider.get(Parser.class);
		
		assertNotNull(parser);
		assertNotEquals(MplDocumentParser.getInstance(), parser);
	}
	
	class MyParser implements Parser<Document>{

		@Override
		public Document parse(InputStream input) throws IOException {
			return null;
		}
		
	}

}

File: source/src/test/kotlin/org/rss/tools/mpl/test/ParserDelegateTest.kt
package org.rss.tools.mpl.test

import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.rss.tools.mpl.builder.DocumentBuilder
import org.rss.tools.mpl.domain.Document
import org.rss.tools.mpl.global.ServiceProvider
import org.rss.tools.mpl.parsing.Parser
import org.rss.tools.mpl.render.DocumentVisitor

import java.io.IOException
import java.io.InputStream

import org.junit.Assert.*
import org.rss.tools.mpl.Main

class ParserDelegateTest {

    @Test
    @Throws(Exception::class)
    internal fun testDefault() {
        val render = ServiceProvider[DocumentVisitor::class.java]
        val parser = ServiceProvider[Parser::class.java]

        assertNotNull(render)
        assertTrue(parser is DocumentBuilder)

        testOverrideParser()
    }

    // Needs to be executed in sequence. Otherwise it breaks the default test
    //	@Test
    @Throws(Exception::class)
    internal fun testOverrideParser() {
        ServiceProvider.register(Parser::class.java, MyParser())
        val parser = ServiceProvider[Parser::class.java]

        assertNotNull(parser)
        assertNotEquals(DocumentBuilder, parser)
    }

    internal inner class MyParser : Parser<Document> {

        @Throws(IOException::class)
        override fun parse(input: InputStream): Document? {
            return null
        }

    }

    companion object {

        @BeforeAll
        fun setup() {
            Main.registerDefaultProviders()
        }
    }

}
